File: web/src/types/assistant.ts
export type AssistantProvider = 'openai' | 'gemini';

export type KgChatClause = {
	id: string;
	label: string;
	burden: number;
	benefit: number;
};

export type KgChatLedger = {
	obligations: number;
	rights: number;
	prohibitions: number;
	burdenWeight: number;
	benefitWeight: number;
	burdenCount: number;
	benefitCount: number;
	usePageRank: boolean;
	topClauses: KgChatClause[];
};

export type AssistantCitation = {
	id: string;
	excerpt: string;
	page?: number;
	paragraph_enum?: number;
};

export type AssistantMessageRole = 'user' | 'assistant';

export type AssistantChatMessage = {
	id: string;
	role: AssistantMessageRole;
	content: string;
	citations?: AssistantCitation[];
	suggestedQuestions?: string[];
	entityHighlights?: Array<{
		label: string;
		key: string;
		color: string;
		softColor: string;
	}>;
};

export type AssistantContextNode = {
	id: string;
	text: string;
	paragraph_enum: number;
	page: number;
};

export type AssistantHistoryMessage = {
	role: AssistantMessageRole;
	content: string;
};

export type AssistantChatRequest = {
	documentId: string;
	question: string;
	provider: AssistantProvider;
	model?: string;
	selectedParagraphId?: string | null;
	paragraphNodes: AssistantContextNode[];
	history: AssistantHistoryMessage[];
	focusNodeId?: string | null;
	focusNodeLabel?: string | null;
	focusNodeKind?: string | null;
	focusParagraphIds?: string[];
	kgLedger?: KgChatLedger | null;
};

export type AssistantChatResponse = {
	answer: string;
	citations: AssistantCitation[];
	suggestedQuestions: string[];
	provider: AssistantProvider;
};

File: server/schemas/assistant.py
from pydantic import BaseModel, Field

from schemas.common import AssistantMessageRole, AssistantProvider


class AssistantParagraphNode(BaseModel):
    id: str
    text: str
    paragraph_enum: int
    page: int


class AssistantHistoryMessage(BaseModel):
    role: AssistantMessageRole
    content: str


class KgChatClause(BaseModel):
    id: str
    label: str
    burden: float
    benefit: float


class KgChatLedger(BaseModel):
    """Deterministic burden/benefit facts for the focused party (already computed
    on the client). The chat explains these; it never recomputes them."""

    obligations: int
    rights: int
    prohibitions: int
    burdenWeight: float
    benefitWeight: float
    burdenCount: int
    benefitCount: int
    usePageRank: bool = True
    topClauses: list[KgChatClause] = Field(default_factory=list)


class AssistantChatRequest(BaseModel):
    documentId: str
    question: str
    provider: AssistantProvider = "openai"
    model: str | None = None
    selectedParagraphId: str | None = None
    paragraphNodes: list[AssistantParagraphNode]
    history: list[AssistantHistoryMessage] = Field(default_factory=list)
    focusNodeId: str | None = None
    focusNodeLabel: str | None = None
    focusNodeKind: str | None = None
    focusParagraphIds: list[str] = Field(default_factory=list)
    kgLedger: KgChatLedger | None = None


class AssistantCitation(BaseModel):
    id: str
    excerpt: str
    page: int | None = None
    paragraph_enum: int | None = None


class AssistantChatResponse(BaseModel):
    answer: str
    citations: list[AssistantCitation]
    suggestedQuestions: list[str]
    provider: AssistantProvider
